hooks/plugin_base.py

```diff
@@ -0,0 +1,133 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+
+import sgtk
+from sgtk.platform.qt import QtCore, QtGui
+
+class PublishPluginBase(sgtk.Hook):
+    """
+    The base class plugin used for all publish plugins. This class is inserted
+    into any loaded publish plugin's class hierarchy by the publisher itself.
+
+    The class implements the default UI state of publish plugins which shows
+    only the description of the plugin. In order to customize the way all
+    plugins are displayed when selected, you can modify this hook.
+
+    To change the way an individual plugin is displayed in the UI, you can
+    override these methods in that plugin. Alternatively, you can make a copy
+    of this hook, configure the new file to be the base class for your plugin,
+    and then make changes to the new file to display the plugin's UI in
+    whatever way you want.
+    """
+
+    def create_settings_widget(self, parent):
```

Comment:
Is this now duplicated code from the app?  If so, any way to call it from the app or have the app use the code from here?

Reply:
Good call. I made changes to keep an instance of the base plugin around and use that whenever we want to generate a default UI for a plugin that doesn't implement the methods. This required a minor hack to inject the current plugin's "description" onto the base plugin instance, but seems like a reasonable tradeoff to not have to duplicate the default description UI logic. Here's the relevant bits: 

* https://github.com/shotgunsoftware/tk-multi-publish2/pull/52/files#diff-6da8c710ddfe37588ac2fc105dda81ccR176
* https://github.com/shotgunsoftware/tk-multi-publish2/pull/52/files#diff-6da8c710ddfe37588ac2fc105dda81ccR338

Let me know what you think @robblau.

Reply:
I think it is fine.  Given that the app is the one providing the hook base class I think it is perfectly fine for the app to have specific knowledge about what it does.